examples/opensensemap/main.c

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright (C) 2018 Inria
+ *               2018 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     examples
+ * @{
+ *
+ * @file
+ * @brief       Example demonstrating how to integrate TTN with openSenseMap
+ *
+ * @author      Alexandre Abadie <alexandre.abadie@inria.fr>
+ * @author      Leandro Lanzieri <leandro.lanzieri@haw-hamburg.de>
+ * @}
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <stdlib.h>
+#include <time.h>
+
+#include "msg.h"
+#include "thread.h"
+#include "fmt.h"
+
+#include "periph/rtc.h"
+
+#include "net/loramac.h"
+#include "semtech_loramac.h"
+#include "lora_serialization.h"
+
+/* Messages are sent every 20s to respect the duty cycle on each channel */
+#define PERIOD              (20U)
+
+#define SENDER_PRIO         (THREAD_PRIORITY_MAIN - 1)
+static kernel_pid_t sender_pid;
+static char sender_stack[THREAD_STACKSIZE_MAIN / 2];
+
+semtech_loramac_t loramac;
+lora_serialization_t serialization;
+
+static uint8_t deveui[LORAMAC_DEVEUI_LEN];
+static uint8_t appeui[LORAMAC_APPEUI_LEN];
+static uint8_t appkey[LORAMAC_APPKEY_LEN];
+
+static void rtc_cb(void *arg)
+{
+    (void) arg;
+    msg_t msg;
+    msg_send(&msg, sender_pid);
+}
+
+static void _prepare_next_alarm(void)
+{
+    struct tm time;
+    rtc_get_time(&time);
+    /* set initial alarm */
+    time.tm_sec += PERIOD;
+    mktime(&time);
+    rtc_set_alarm(&time, rtc_cb, NULL);
+}
+
+static void _send_message(void)
+{
+    printf("Sending data\n");
+
+    /* Reset serialization descriptor */
+    lora_serialization_reset(&serialization);
+
+    /* Write data to serialization. Replace with your sensors measurements.
+     * Keep in mind that the order in which the data is written into the
+     * serialization needs to match the decoders order.
+     */
+    lora_serialization_write_temperature(&serialization, 25.3);
+    lora_serialization_write_uint8(&serialization, 82);
+    lora_serialization_write_humidity(&serialization, 50.2);
```

Comment:
I see that this is very similar to the lorawan example. My thoughts are however that this example could be designed more towards presenting a clean, extensible interface to the user. Personally I like the structure of e.g. the gcoap and dtls-echo examples - they put a simple interface to the program in main.c. Then, if people want to change other aspects of how the program responds, they can go into the other .c files.

I would suggest finding a program structure so that in main.c they can do something like:
- add measurements to be sent in a declarative manner in a simple macro or function (for Sensebox this could also add the drivers at compile time, not sure whether this would require a separate macro/function or whether it could be done automatically). They could perhaps specify that the application sends dummy data (as above), or direct from sensor, or from a custom function
- change the frequency of data sending

Then in another file (loramac_send.c?) 
- asyncronous data sending stuff incl thread definition perhaps
- arrays storing keys

Basically the philosophy I'm suggesting is to present a really simple API to a user in main.c which allows them to quickly, easily and flexibly write data sending applications according to their own hardware setup. As easy as possible. E.g. as little as possible above just plugging together their sensebox and loading this application.